File: src/leetcode/js/reverse-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseListWithIterative = function(head) {
    let prevNode = null
    let currNode = head

    while (currNode !== null) {
        const nextNode = currNode.next
        currNode.next = prevNode
        prevNode = currNode
        currNode = nextNode
    }

    return prevNode
};

var reverseListWithRecursive = function(head) {
    if (head === null || head.next === null) {
        return head
    }

    const reversedSublist = reverseList(head.next)
    head.next.next = head
    head.next = null

    return reversedSublist
};

File: src/leetcode/java/ReverseLinkedList.java
package leetcode.java;

public class ReverseLinkedList {
    public static class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public ListNode reverseListWithIterative(ListNode head) {
        ListNode prevNode = null;
        ListNode currNode = head;

        while (currNode != null) {
            ListNode nextNode = currNode.next;
            currNode.next = prevNode;
            prevNode = currNode;
            currNode = nextNode;
        }

        return prevNode;
    }

    public ListNode reverseListWithRecursive(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode reversedSublist = reverseListWithRecursive(head.next);
        head.next.next = head;
        head.next = null;

        return reversedSublist;
    }
}
